### PR TITLE
Fix `DynamicThreadPool` may never be destructed, causing tiflash node to fail to exit normally

### LINKED ISSUE #7857: DynamicThreadPool may never be destructed, causing tiflash node to fail to exit normally
```
https://github.com/pingcap/tiflash/blob/dfb7e35c14d64d95ae64c1f68ffaab112af46c79/dbms/src/Common/DynamicThreadPool.cpp#L37-L38
DynamicThreadPool::~DynamicThreadPool()
{
    ...
    // TODO: maybe use a latch is more elegant.
    while (alive_dynamic_threads.load() != 0)
        std::this_thread::sleep_for(std::chrono::milliseconds(1));
}
```
As shown above, the destructor of `DynamicThreadPool` will wait for `alive_dynamic_threads` to return to 0 before exiting.

```
https://github.com/pingcap/tiflash/blob/dfb7e35c14d64d95ae64c1f68ffaab112af46c79/dbms/src/Common/DynamicThreadPool.cpp#L91-L96
void DynamicThreadPool::scheduledToNewDynamicThread(TaskPtr & task)
{
    alive_dynamic_threads.fetch_add(1);
    std::thread t = ThreadFactory::newThread(false, "DynamicThread", &DynamicThreadPool::dynamicWork, this, std::move(task));
    t.detach();
}

https://github.com/pingcap/tiflash/blob/dfb7e35c14d64d95ae64c1f68ffaab112af46c79/dbms/src/Common/DynamicThreadPool.cpp#L147
void DynamicThreadPool::dynamicWork(TaskPtr initial_task)
{
   ...
    alive_dynamic_threads.fetch_sub(1);
}
```
As shown above, `alive_dynamic_threads` is incremented before `ThreadFactory::newThread`, and decremented in `DynamicThreadPool::dynamicWork`.
If `ThreadFactory::newThread` throws an error, such as thread `constructor failed: Resource temporarily unavailable`, then `alive_dynamic_threads` will never return to 0, which means that `DynamicThreadPool` will never be destructed.

### PR DESCRIPTION
**Pull request by SeaRise**
### What problem does this PR solve?

Issue Number: close #7857

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Review by guo-shaoge — APPROVED** [review]

File: dbms/src/Common/DynamicThreadPool.cpp [hunk 1]
```diff
@@ -90,11 +98,25 @@ bool DynamicThreadPool::scheduledToExistedDynamicThread(TaskPtr & task)
...
-    std::thread t = ThreadFactory::newThread(false, "DynamicThread", &DynamicThreadPool::dynamicWork, this, std::move(task));
+    std::thread t = newDynamcThread(task);
     t.detach();
 }
 
+std::thread DynamicThreadPool::newDynamcThread(TaskPtr & task)
+{
+    try
+    {
+        alive_dynamic_threads.fetch_add(1);
```
Comment (SeaRise) [inline dbms/src/Common/DynamicThreadPool.cpp:109]:
```suggestion
    alive_dynamic_threads.fetch_add(1);
    try
    {
```

Comment (guo-shaoge) [reply dbms/src/Common/DynamicThreadPool.cpp]: Why change ?

**Review by Lloyd-Pottiger — APPROVED** [review]

**Review by windtalker — APPROVED** [review]
LGTM

**Pull request merged**